File: clients/cpp/model/BuildingAction.cpp
#include "BuildingAction.hpp"

namespace model {

BuildingAction::BuildingAction(int planet, std::optional<model::BuildingType> buildingType) : planet(planet), buildingType(buildingType) { }

// Read BuildingAction from input stream
BuildingAction BuildingAction::readFrom(InputStream& stream) {
    int planet = stream.readInt();
    std::optional<model::BuildingType> buildingType = std::optional<model::BuildingType>();
    if (stream.readBool()) {
        model::BuildingType buildingTypeValue = readBuildingType(stream);
        buildingType.emplace(buildingTypeValue);
    }
    return BuildingAction(planet, buildingType);
}

// Write BuildingAction to output stream
void BuildingAction::writeTo(OutputStream& stream) const {
    stream.write(planet);
    if (buildingType) {
        stream.write(true);
        const model::BuildingType& buildingTypeValue = *buildingType;
        stream.write((int)(buildingTypeValue));
    } else {
        stream.write(false);
    }
}

// Get string representation of BuildingAction
std::string BuildingAction::toString() const {
    std::stringstream ss;
    ss << "BuildingAction { ";
    ss << "planet: ";
    ss << planet;
    ss << ", ";
    ss << "buildingType: ";
    if (buildingType) {
        const model::BuildingType& buildingTypeValue = *buildingType;
        ss << buildingTypeToString(buildingTypeValue);
    } else {
        ss << "none";
    }
    ss << " }";
    return ss.str();
}

}

File: clients/ruby/model/building_action.rb
require './model/building_type'

module Model

# TODO - Document
class BuildingAction
    # TODO - Document
    attr_accessor :planet
    # TODO - Document
    attr_accessor :building_type

    def initialize(planet, building_type)
        @planet = planet
        @building_type = building_type
    end

    # Read BuildingAction from input stream
    def self.read_from(stream)
        planet = stream.read_int()
        if stream.read_bool()
            building_type = Model::BuildingType.read_from(stream)
        else
            building_type = nil
        end
        BuildingAction.new(planet, building_type)
    end

    # Write BuildingAction to output stream
    def write_to(stream)
        stream.write_int(@planet)
        if @building_type.nil?
            stream.write_bool(false)
        else
            stream.write_bool(true)
            stream.write_int(@building_type)
        end
    end

    def to_s
        string_result = "BuildingAction { "
        string_result += "planet: "
        string_result += @planet.to_s
        string_result += ", "
        string_result += "building_type: "
        if @building_type.nil?
            string_result += "nil"
        else
            string_result += BuildingType.to_s(@building_type)
        end
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end